manager/src/main/java/com/redhat/service/bridge/manager/BridgeWorker.java

```diff
@@ -0,0 +1,208 @@
+
+        routePreparingWorker(preparingWorker, bridge);
+    }
+
+    @Transactional
+    @ConsumeEvent(value = CREATE_KAFKA_TOPIC, blocking = true)
+    public void step1_createKafkaTopic(Bridge bridge) {
+        try {
+            rhoasService.createTopicAndGrantAccessFor(internalKafkaConfigurationProvider.getBridgeTopicName(bridge), RhoasTopicAccessType.CONSUMER_AND_PRODUCER);
+        } catch (RuntimeException e) {
+            LOGGER.error("Failed to create the topic. The retry will be performed by the reschedule loop.", e);
+            updatePreparingWorker(bridge.getId(), BridgeWorkerStatus.FAILURE);
+            return;
+        }
+
+        PreparingWorker preparingWorker = preparingWorkerDAO.findByEntityId(bridge.getId()).get(0); // there should be no parallel workers for this resource.
+        preparingWorkerDAO.deleteById(preparingWorker.getId());
+
+        // The bridge is ready to be provisioned by the shard.
+        bridge.setStatus(ManagedEntityStatus.PREPARING);
+        bridge.setDesiredStatus(ManagedEntityStatus.PROVISIONING);
+        transact(bridge);
+    }
+
+    @Transactional
+    @ConsumeEvent(value = DELETE_KAFKA_TOPIC, blocking = true)
+    public void step1_deleteKafkaTopic(Bridge bridge) {
+        try {
+            rhoasService.deleteTopicAndRevokeAccessFor(internalKafkaConfigurationProvider.getBridgeTopicName(bridge), RhoasTopicAccessType.CONSUMER_AND_PRODUCER);
+        } catch (RuntimeException e) {
+            LOGGER.error("Failed to create the topic. The retry will be performed by the reschedule loop.", e);
+            updatePreparingWorker(bridge.getId(), BridgeWorkerStatus.FAILURE);
+            return;
+        }
+
+        PreparingWorker preparingWorker = preparingWorkerDAO.findByEntityId(bridge.getId()).get(0); // there should be no parallel workers for this resource.
+        preparingWorkerDAO.deleteById(preparingWorker.getId());
+
+        // The bridge is ready to be provisioned by the shard.
+        bridge.setStatus(ManagedEntityStatus.DEPROVISION);
+        bridge.setDesiredStatus(ManagedEntityStatus.DELETING);
+        transact(bridge);
+    }
+
+    @Override
+    @Scheduled(every = "30s", concurrentExecution = Scheduled.ConcurrentExecution.SKIP)
+    public void reschedule() {
+        LOGGER.info("rescheduled");
+        List<PreparingWorker> preparingWorkers = preparingWorkerDAO.findByWorkerIdAndStatusAndType(workerIdProvider.getWorkerId(), BridgeWorkerStatus.FAILURE.toString(), "BRIDGE");
+        for (PreparingWorker preparingWorker : preparingWorkers) {
+            Bridge bridge = bridgeDAO.findById(preparingWorker.getEntityId());
+            routePreparingWorker(preparingWorker, bridge);
+        }
+    }
+
+    /**
+     * Every 5 minutes it will retrieve all the instances that have not been updated for more than 30 minutes.
+     */
+    @Override
+    @Scheduled(every = "5m", concurrentExecution = Scheduled.ConcurrentExecution.SKIP)
+    public void discoverOrphanWorkers() {
```

Comment:
If I step back, this feels like we're trying to achieve what Kafka could give us: the ability to have multiple workers process a single job queue. 

We have a hard dependency on Kafka for OpenBridge. If we were to consider a work queue for each customer account and then the workers simply become kafka consumers, does that make our life simpler?

At the minute we have an artificial 5 min wait period should a worker get rescheduled (we can of course reduce this through configuration). If instead we didn't commit that the offset for the event on Kafka has been processed, we have the ability to simply bring up another consumer straight away and continue processing.

Reply:
I tend to agree; we are inventing our own durable and resilient state machine.

This is in essence the same pattern that we are achieving with our use of the DAO and VertX Event Bus.

It's definitely more things to consider on Monday.